File: i2c/src/17/17.c
#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include <math.h>

// I2C speed mode definitions
typedef enum {
    I2C_STANDARD_MODE = 0,    // 100 kHz
    I2C_FAST_MODE,            // 400 kHz
    I2C_FAST_MODE_PLUS,       // 1 MHz
    I2C_HIGH_SPEED_MODE       // 3.4 MHz
} i2c_speed_mode_t;

// Capacitance limits (in pF)
static const uint16_t I2C_CAP_LIMITS[] = {
    400,  // Standard Mode
    400,  // Fast Mode
    550,  // Fast Mode Plus
    100   // High-Speed Mode
};

// Speed frequencies (in Hz)
static const uint32_t I2C_FREQUENCIES[] = {
    100000,   // Standard Mode
    400000,   // Fast Mode
    1000000,  // Fast Mode Plus
    3400000   // High-Speed Mode
};

// Structure to hold bus capacitance components
typedef struct {
    float trace_length_m;        // PCB trace length in meters
    float trace_cap_per_m;       // Capacitance per meter (pF/m)
    uint8_t num_devices;         // Number of I2C devices
    float device_cap_pf;         // Average device input capacitance (pF)
    uint8_t num_connectors;      // Number of connectors
    float connector_cap_pf;      // Capacitance per connector (pF)
    float parasitic_cap_pf;      // Estimated parasitic capacitance (pF)
} i2c_bus_components_t;

// Calculate total bus capacitance
float calculate_bus_capacitance(const i2c_bus_components_t* components) {
    float trace_cap = components->trace_length_m * components->trace_cap_per_m;
    float device_cap = components->num_devices * components->device_cap_pf;
    float connector_cap = components->num_connectors * components->connector_cap_pf;
    
    return trace_cap + device_cap + connector_cap + components->parasitic_cap_pf;
}

// Validate if capacitance is within limits for given mode
bool validate_capacitance(float total_cap_pf, i2c_speed_mode_t mode) {
    return total_cap_pf <= I2C_CAP_LIMITS[mode];
}

// Calculate required pull-up resistor value
// Formula: R = t_rise / (0.8473 * C_bus)
// where t_rise is typically 1000ns for Standard/Fast, 300ns for Fast+, 80ns for HS
float calculate_pullup_resistor(float bus_cap_pf, i2c_speed_mode_t mode) {
    float t_rise_ns;
    
    switch(mode) {
        case I2C_STANDARD_MODE:
        case I2C_FAST_MODE:
            t_rise_ns = 1000.0f;  // 1000ns max rise time
            break;
        case I2C_FAST_MODE_PLUS:
            t_rise_ns = 300.0f;   // 300ns max rise time
            break;
        case I2C_HIGH_SPEED_MODE:
            t_rise_ns = 80.0f;    // 80ns max rise time
            break;
        default:
            t_rise_ns = 1000.0f;
    }
    
    // Convert pF to F: pF * 1e-12
    float bus_cap_f = bus_cap_pf * 1e-12f;
    float t_rise_s = t_rise_ns * 1e-9f;
    
    // Minimum resistor: R_min = t_rise / (0.8473 * C)
    float r_min = t_rise_s / (0.8473f * bus_cap_f);
    
    return r_min;
}

// Calculate maximum bus capacitance for given pull-up resistor
float calculate_max_capacitance(float pullup_ohms, i2c_speed_mode_t mode) {
    float t_rise_ns;
    
    switch(mode) {
        case I2C_STANDARD_MODE:
        case I2C_FAST_MODE:
            t_rise_ns = 1000.0f;
            break;
        case I2C_FAST_MODE_PLUS:
            t_rise_ns = 300.0f;
            break;
        case I2C_HIGH_SPEED_MODE:
            t_rise_ns = 80.0f;
            break;
        default:
            t_rise_ns = 1000.0f;
    }
    
    float t_rise_s = t_rise_ns * 1e-9f;
    
    // C_max = t_rise / (0.8473 * R)
    float c_max_f = t_rise_s / (0.8473f * pullup_ohms);
    
    // Convert to pF
    return c_max_f * 1e12f;
}

// Print detailed capacitance report
void print_capacitance_report(const i2c_bus_components_t* components, 
                               i2c_speed_mode_t mode) {
    const char* mode_names[] = {
        "Standard Mode (100 kHz)",
        "Fast Mode (400 kHz)",
        "Fast Mode Plus (1 MHz)",
        "High-Speed Mode (3.4 MHz)"
    };
    
    float total_cap = calculate_bus_capacitance(components);
    bool is_valid = validate_capacitance(total_cap, mode);
    float min_pullup = calculate_pullup_resistor(total_cap, mode);
    
    printf("\n========== I2C Bus Capacitance Report ==========\n");
    printf("Target Mode: %s\n\n", mode_names[mode]);
    
    printf("Capacitance Components:\n");
    printf("  PCB Trace: %.1f cm × %.1f pF/m = %.2f pF\n",
           components->trace_length_m * 100,
           components->trace_cap_per_m,
           components->trace_length_m * components->trace_cap_per_m);
    printf("  Devices: %d × %.1f pF = %.2f pF\n",
           components->num_devices,
           components->device_cap_pf,
           components->num_devices * components->device_cap_pf);
    printf("  Connectors: %d × %.1f pF = %.2f pF\n",
           components->num_connectors,
           components->connector_cap_pf,
           components->num_connectors * components->connector_cap_pf);
    printf("  Parasitic: %.2f pF\n", components->parasitic_cap_pf);
    printf("  -----------------------------------\n");
    printf("  TOTAL: %.2f pF\n\n", total_cap);
    
    printf("Validation:\n");
    printf("  Maximum allowed: %d pF\n", I2C_CAP_LIMITS[mode]);
    printf("  Status: %s\n", is_valid ? "PASS ✓" : "FAIL ✗");
    printf("  Margin: %.2f pF (%.1f%%)\n\n",
           I2C_CAP_LIMITS[mode] - total_cap,
           ((I2C_CAP_LIMITS[mode] - total_cap) / I2C_CAP_LIMITS[mode]) * 100);
    
    printf("Pull-up Resistor:\n");
    printf("  Minimum required: %.0f Ω (%.2f kΩ)\n",
           min_pullup, min_pullup / 1000.0f);
    printf("  Recommended: %.2f kΩ (standard value)\n",
           ceil(min_pullup / 1000.0f));
    printf("===============================================\n\n");
}

int main() {
    // Example 1: Typical sensor system
    i2c_bus_components_t sensor_bus = {
        .trace_length_m = 0.15f,      // 15 cm trace
        .trace_cap_per_m = 35.0f,     // 35 pF/m typical
        .num_devices = 4,             // 4 sensors
        .device_cap_pf = 6.0f,        // 6 pF per device
        .num_connectors = 1,          // 1 connector
        .connector_cap_pf = 4.0f,     // 4 pF connector
        .parasitic_cap_pf = 8.0f      // 8 pF estimated parasitic
    };
    
    print_capacitance_report(&sensor_bus, I2C_FAST_MODE);
    
    // Example 2: Long distance system
    i2c_bus_components_t long_bus = {
        .trace_length_m = 1.0f,       // 1 meter trace
        .trace_cap_per_m = 40.0f,     // 40 pF/m
        .num_devices = 8,             // 8 devices
        .device_cap_pf = 7.0f,        // 7 pF per device
        .num_connectors = 3,          // 3 connectors
        .connector_cap_pf = 5.0f,     // 5 pF per connector
        .parasitic_cap_pf = 15.0f     // 15 pF estimated parasitic
    };
    
    print_capacitance_report(&long_bus, I2C_STANDARD_MODE);
    
    // Demonstrate maximum capacitance calculation
    printf("Maximum Capacitance for 4.7kΩ Pull-up:\n");
    for (int mode = 0; mode < 4; mode++) {
        float max_cap = calculate_max_capacitance(4700.0f, (i2c_speed_mode_t)mode);
        printf("  %s: %.2f pF\n", 
               mode == 0 ? "Standard" : mode == 1 ? "Fast" : 
               mode == 2 ? "Fast+" : "High-Speed",
               max_cap);
    }
    
    return 0;
}

File: I2C/src/17/17.rs
use std::fmt;

/// I2C speed modes with associated specifications
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum SpeedMode {
    Standard,     // 100 kHz
    Fast,         // 400 kHz
    FastPlus,     // 1 MHz
    HighSpeed,    // 3.4 MHz
}

impl SpeedMode {
    /// Maximum bus capacitance in picofarads
    pub fn max_capacitance(&self) -> f32 {
        match self {
            SpeedMode::Standard => 400.0,
            SpeedMode::Fast => 400.0,
            SpeedMode::FastPlus => 550.0,
            SpeedMode::HighSpeed => 100.0,
        }
    }
    
    /// Operating frequency in Hz
    pub fn frequency(&self) -> u32 {
        match self {
            SpeedMode::Standard => 100_000,
            SpeedMode::Fast => 400_000,
            SpeedMode::FastPlus => 1_000_000,
            SpeedMode::HighSpeed => 3_400_000,
        }
    }
    
    /// Maximum rise time in nanoseconds
    pub fn max_rise_time_ns(&self) -> f32 {
        match self {
            SpeedMode::Standard | SpeedMode::Fast => 1000.0,
            SpeedMode::FastPlus => 300.0,
            SpeedMode::HighSpeed => 80.0,
        }
    }
}

impl fmt::Display for SpeedMode {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            SpeedMode::Standard => write!(f, "Standard Mode (100 kHz)"),
            SpeedMode::Fast => write!(f, "Fast Mode (400 kHz)"),
            SpeedMode::FastPlus => write!(f, "Fast Mode Plus (1 MHz)"),
            SpeedMode::HighSpeed => write!(f, "High-Speed Mode (3.4 MHz)"),
        }
    }
}

/// Components contributing to total bus capacitance
#[derive(Debug, Clone)]
pub struct BusComponents {
    pub trace_length_m: f32,        // PCB trace length in meters
    pub trace_cap_per_m: f32,       // Capacitance per meter (pF/m)
    pub num_devices: u8,            // Number of I2C devices
    pub device_cap_pf: f32,         // Average device input capacitance (pF)
    pub num_connectors: u8,         // Number of connectors
    pub connector_cap_pf: f32,      // Capacitance per connector (pF)
    pub parasitic_cap_pf: f32,      // Estimated parasitic capacitance (pF)
}

impl BusComponents {
    /// Create a new bus components configuration
    pub fn new() -> Self {
        Self {
            trace_length_m: 0.0,
            trace_cap_per_m: 35.0,  // Typical PCB trace
            num_devices: 0,
            device_cap_pf: 6.0,     // Typical I2C device
            num_connectors: 0,
            connector_cap_pf: 4.0,  // Typical connector
            parasitic_cap_pf: 5.0,  // Conservative estimate
        }
    }
    
    /// Builder method: set trace length
    pub fn with_trace(mut self, length_m: f32, cap_per_m: f32) -> Self {
        self.trace_length_m = length_m;
        self.trace_cap_per_m = cap_per_m;
        self
    }
    
    /// Builder method: set devices
    pub fn with_devices(mut self, count: u8, cap_pf: f32) -> Self {
        self.num_devices = count;
        self.device_cap_pf = cap_pf;
        self
    }
    
    /// Builder method: set connectors
    pub fn with_connectors(mut self, count: u8, cap_pf: f32) -> Self {
        self.num_connectors = count;
        self.connector_cap_pf = cap_pf;
        self
    }
    
    /// Builder method: set parasitic capacitance
    pub fn with_parasitic(mut self, cap_pf: f32) -> Self {
        self.parasitic_cap_pf = cap_pf;
        self
    }
}

/// Result of capacitance analysis
#[derive(Debug)]
pub struct CapacitanceAnalysis {
    pub total_capacitance: f32,
    pub trace_contribution: f32,
    pub device_contribution: f32,
    pub connector_contribution: f32,
    pub parasitic_contribution: f32,
    pub is_valid: bool,
    pub margin_pf: f32,
    pub margin_percent: f32,
    pub min_pullup_ohms: f32,
}

/// I2C bus capacitance calculator
pub struct CapacitanceCalculator;

impl CapacitanceCalculator {
    /// Calculate total bus capacitance from components
    pub fn calculate(components: &BusComponents) -> f32 {
        let trace = components.trace_length_m * components.trace_cap_per_m;
        let devices = components.num_devices as f32 * components.device_cap_pf;
        let connectors = components.num_connectors as f32 * components.connector_cap_pf;
        
        trace + devices + connectors + components.parasitic_cap_pf
    }
    
    /// Perform full analysis for a given speed mode
    pub fn analyze(components: &BusComponents, mode: SpeedMode) -> CapacitanceAnalysis {
        let trace_contrib = components.trace_length_m * components.trace_cap_per_m;
        let device_contrib = components.num_devices as f32 * components.device_cap_pf;
        let connector_contrib = components.num_connectors as f32 * components.connector_cap_pf;
        let parasitic_contrib = components.parasitic_cap_pf;
        
        let total = trace_contrib + device_contrib + connector_contrib + parasitic_contrib;
        let max_allowed = mode.max_capacitance();
        let is_valid = total <= max_allowed;
        let margin = max_allowed - total;
        let margin_pct = (margin / max_allowed) * 100.0;
        
        let min_pullup = Self::calculate_min_pullup(total, mode);
        
        CapacitanceAnalysis {
            total_capacitance: total,
            trace_contribution: trace_contrib,
            device_contribution: device_contrib,
            connector_contribution: connector_contrib,
            parasitic_contribution: parasitic_contrib,
            is_valid,
            margin_pf: margin,
            margin_percent: margin_pct,
            min_pullup_ohms: min_pullup,
        }
    }
    
    /// Calculate minimum pull-up resistor value
    /// Formula: R_min = t_rise / (0.8473 * C_bus)
    pub fn calculate_min_pullup(bus_cap_pf: f32, mode: SpeedMode) -> f32 {
        let t_rise_s = mode.max_rise_time_ns() * 1e-9;
        let bus_cap_f = bus_cap_pf * 1e-12;
        
        t_rise_s / (0.8473 * bus_cap_f)
    }
    
    /// Calculate maximum capacitance for given pull-up resistor
    pub fn calculate_max_cap(pullup_ohms: f32, mode: SpeedMode) -> f32 {
        let t_rise_s = mode.max_rise_time_ns() * 1e-9;
        let c_max_f = t_rise_s / (0.8473 * pullup_ohms);
        
        c_max_f * 1e12  // Convert to pF
    }
    
    /// Find highest compatible speed mode for given capacitance
    pub fn find_max_speed(total_cap_pf: f32) -> Option<SpeedMode> {
        let modes = [
            SpeedMode::HighSpeed,
            SpeedMode::FastPlus,
            SpeedMode::Fast,
            SpeedMode::Standard,
        ];
        
        modes.iter()
            .find(|mode| total_cap_pf <= mode.max_capacitance())
            .copied()
    }
}

impl CapacitanceAnalysis {
    /// Print detailed report
    pub fn print_report(&self, mode: SpeedMode) {
        println!("\n========== I2C Bus Capacitance Report ==========");
        println!("Target Mode: {}", mode);
        println!();
        
        println!("Capacitance Breakdown:");
        println!("  PCB Trace:    {:6.2} pF", self.trace_contribution);
        println!("  Devices:      {:6.2} pF", self.device_contribution);
        println!("  Connectors:   {:6.2} pF", self.connector_contribution);
        println!("  Parasitic:    {:6.2} pF", self.parasitic_contribution);
        println!("  {}", "-".repeat(25));
        println!("  TOTAL:        {:6.2} pF", self.total_capacitance);
        println!();
        
        println!("Validation:");
        println!("  Maximum allowed:  {:.0} pF", mode.max_capacitance());
        println!("  Status:           {}", if self.is_valid { "PASS ✓" } else { "FAIL ✗" });
        println!("  Margin:           {:.2} pF ({:.1}%)", self.margin_pf, self.margin_percent);
        println!();
        
        println!("Pull-up Resistor:");
        println!("  Minimum required: {:.0} Ω ({:.2} kΩ)", 
                 self.min_pullup_ohms, 
                 self.min_pullup_ohms / 1000.0);
        
        // Suggest standard resistor values
        let standard_values = [1000.0, 1500.0, 2200.0, 3300.0, 4700.0, 10000.0];
        if let Some(&recommended) = standard_values.iter()
            .find(|&&val| val >= self.min_pullup_ohms) {
            println!("  Recommended:      {:.1} kΩ (standard value)", recommended / 1000.0);
        }
        
        println!("===============================================\n");
    }
}

fn main() {
    println!("I2C Bus Capacitance Analysis Tool\n");
    
    // Example 1: Compact sensor board
    let sensor_board = BusComponents::new()
        .with_trace(0.12, 35.0)      // 12 cm trace, 35 pF/m
        .with_devices(3, 5.5)        // 3 sensors, 5.5 pF each
        .with_connectors(0, 0.0)     // No connectors
        .with_parasitic(6.0);        // 6 pF parasitic
    
    let analysis = CapacitanceCalculator::analyze(&sensor_board, SpeedMode::Fast);
    analysis.print_report(SpeedMode::Fast);
    
    // Example 2: Extended system with multiple devices
    let extended_system = BusComponents::new()
        .with_trace(0.50, 40.0)      // 50 cm trace, 40 pF/m
        .with_devices(6, 7.0)        // 6 devices, 7 pF each
        .with_connectors(2, 5.0)     // 2 connectors, 5 pF each
        .with_parasitic(12.0);       // 12 pF parasitic
    
    let total_cap = CapacitanceCalculator::calculate(&extended_system);
    println!("Extended System Total Capacitance: {:.2} pF", total_cap);
    
    if let Some(max_mode) = CapacitanceCalculator::find_max_speed(total_cap) {
        println!("Maximum compatible speed: {}\n", max_mode);
        let analysis = CapacitanceCalculator::analyze(&extended_system, max_mode);
        analysis.print_report(max_mode);
    }
    
    // Example 3: Pull-up resistor analysis
    println!("Maximum Capacitance for Common Pull-up Values:");
    let pullup_values = [2200.0, 3300.0, 4700.0, 10000.0];
    
    for &pullup in &pullup_values {
        println!("\n{:.1} kΩ Pull-up:", pullup / 1000.0);
        for mode in [SpeedMode::Standard, SpeedMode::Fast, SpeedMode::FastPlus, SpeedMode::HighSpeed] {
            let max_cap = CapacitanceCalculator::calculate_max_cap(pullup, mode);
            let spec_limit = mode.max_capacitance();
            let effective = max_cap.min(spec_limit);
            println!("  {:<25} {:.1} pF (limit: {:.0} pF)", 
                     format!("{}", mode), effective, spec_limit);
        }
    }
}